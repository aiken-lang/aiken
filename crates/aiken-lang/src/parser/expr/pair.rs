use crate::{
    builtins::{PAIR, PRELUDE},
    expr::UntypedExpr,
    parser::{error::ParseError, token::Token},
};
use chumsky::prelude::*;

pub fn parser(
    r: Recursive<'_, Token, UntypedExpr, ParseError>,
) -> impl Parser<Token, UntypedExpr, Error = ParseError> + '_ {
    select! {Token::Name { name } if name == PRELUDE => name}
        .then_ignore(just(Token::Dot))
        .or_not()
        .then_ignore(select! {Token::UpName { name } if name == PAIR => name})
        .ignore_then(
            r.clone()
                .separated_by(just(Token::Comma))
                .exactly(2)
                .allow_trailing()
                .delimited_by(
                    choice((just(Token::LeftParen), just(Token::NewLineLeftParen))),
                    just(Token::RightParen),
                )
                .map_with_span(|elems, location| UntypedExpr::Pair {
                    location,
                    fst: elems
                        .first()
                        .expect("Pair should have exactly 2 elements")
                        .to_owned()
                        .into(),
                    snd: elems
                        .last()
                        .expect("Pair should have exactly 2 elements")
                        .to_owned()
                        .into(),
                }),
        )
}

#[cfg(test)]
mod tests {
    use crate::assert_expr;

    #[test]
    fn basic_pair() {
        assert_expr!(r#"Pair(1, 2)"#);
    }

    #[test]
    fn pair_from_prelude() {
        assert_expr!(r#"aiken.Pair(1, 2)"#);
    }
}
