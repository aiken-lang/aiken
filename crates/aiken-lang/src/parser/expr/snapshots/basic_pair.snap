---
source: crates/aiken-lang/src/parser/expr/pair.rs
description: "Code:\n\nPair(1, 2)"
---
Pair {
    location: 4..10,
    fst: UInt {
        location: 5..6,
        value: "1",
        base: Decimal {
            numeric_underscore: false,
        },
    },
    snd: UInt {
        location: 8..9,
        value: "2",
        base: Decimal {
            numeric_underscore: false,
        },
    },
}
